lib/linked_list.rb

```diff
@@ -18,129 +21,272 @@ def initialize
+        @head = current.next
+      end
+
+      # if value is anywhere else in the list
+      while (current != nil)
+        if current.next != nil && current.next.data == value
+          current.next = current.next.next
+        end
+        current = current.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      if @head == nil || @head.next == nil
+        return @head
+      end
+
+      current = @head
+      previous = nil
+
+      while (current != nil)
+        temp = current.next
+        current.next = previous
+        previous = current
+        current = temp
+      end
+      @head = previous
+      return @head.data
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
+    # Time Complexity:
     # Space Complexity
     def find_middle_value
       raise NotImplementedError
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
+    # Time Complexity:
     # Space Complexity
     def find_nth_from_end(n)
-      raise NotImplementedError
+      # find length
+      if @head == nil
+        return @head
+      end
+
+      current = @head
+      count = 0
+      while (current != nil)
+        current = current.next
+        count += 1
+      end
+
+      # find requested value
+      current = @head
+      value = 0
+      while(current != nil)
+        if (value == count - n - 1)
+          return current.data
+        end
+        value += 1
+        current = current.next
+      end
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
+    # Time Complexity:
     # Space Complexity
     def has_cycle
       raise NotImplementedError
     end
 
 
-    # Additional Exercises 
+    # Additional Exercises
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity:  
+    # Time Complexity:
     # Space Complexity
     def get_first
-      raise NotImplementedError
+      if @head == nil
+        return nil
+      else
+        return @head.data
+      end
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
+    # Time Complexity:
     # Space Complexity
     def add_last(value)
-      raise NotImplementedError
+      new_node = Node.new(value)
+      if @head == nil
+        @head = new_node
+        return @head.data
+      end
+
+      current = @head
+      while (current != nil)

```

Comment:
I think you could make the loop `until current.next.nil?` and put the if statement after the loop.